zebra-chain/src/serialization/zcash_serialize.rs

```diff
@@ -27,6 +27,12 @@ pub trait ZcashSerialize: Sized {
         self.zcash_serialize(&mut data)?;
         Ok(data)
     }
+
+    /// Get the size of `self` by using the `zcash_serialize_to_vec()` helper
+    /// to get a vector of bytes, then just apply `len()` on it.
+    fn zcash_serialized_size(&self) -> Result<usize, io::Error> {
+        Ok(self.zcash_serialize_to_vec()?.len())
```

Comment:
Efficiency: this code allocates RAM for a vector, writes data to it, then drops the vector without using the data.

Can you try moving the code from `body_length` into `zcash_serialized_size`?
Then you can make `body_length` call `zcash_serialized_size`.

`body_length` is more efficient, because it doesn't actually allocate any RAM, or write any data:
https://github.com/ZcashFoundation/zebra/blob/de589f6b3113af62c98e327edf969f2ef176a152/zebra-network/src/protocol/external/codec.rs#L176-L183